src/server/lock_manager/mod.rs

```diff
@@ -17,13 +17,22 @@ use crate::server::resolve::StoreAddrResolver;
 use crate::server::{Error, Result};
 use crate::storage::txn::{execute_callback, ProcessResult};
 use crate::storage::{lock_manager::Lock, LockManager as LockManagerTrait, StorageCb};
-use pd_client::RpcClient;
+use pd_client::PdClient;
+use spin::Mutex;
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 use std::thread::JoinHandle;
+use tikv_util::collections::HashSet;
 use tikv_util::security::SecurityManager;
 use tikv_util::worker::FutureWorker;
 
+const DETECTED_SLOTS_NUM: usize = 128;
+
+#[inline]
+fn detected_slot_idx(txn_ts: u64) -> usize {
+    txn_ts as usize & (DETECTED_SLOTS_NUM - 1)
```

Comment:
I'm afraid there can be a lot of contension. The low bits of a timestamp is the logical timestamp, which is typically small. Maybe better to do a xorshift first.

Reply:
Good idea.

Reply:
I use default hasher here.